src/perception/aruco_processing.cpp

```diff
@@ -0,0 +1,136 @@
+#include "aruco_detect.hpp"
+
+#include "filter.hpp"
+
+// TODO: add cache logic to filter out false positives
+
+/**
+ * Detect fiducials from raw image using OpenCV and calculate their screen space centers.
+ * Maintain the immediate buffer - holds all fiducials seen currently on screen.
+ * Later camera space pose information is filled in when we receive point cloud data.
+ *
+ * @param msg
+ */
+void FiducialsNode::imageCallback(sensor_msgs::ImageConstPtr const& msg) {
+    if (!mEnableDetections) return;
+
+    if (mIsVerbose) {
+        ROS_INFO("Got image %d", msg->header.seq);
+    }
+
+    try {
+        mCvPtr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
+
+        // Detect the fiducial vertices in screen space and their respective ids
+        cv::aruco::detectMarkers(mCvPtr->image, mDictionary, mCornersCache, mIdsCache, mDetectorParams);
+
+        // Save fiducials currently on screen to the immediate buffer
+        for (size_t i = 0; i < mIdsCache.size(); ++i) {
+            int id = mIdsCache[i];
+            cv::Point2f imageCenter = std::accumulate(mCornersCache[i].begin(), mCornersCache[i].end(), cv::Point2f{}) / 4.0f;
+            mImmediateFiducials[id].id = id;
+            mImmediateFiducials[id].imageCenter = imageCenter;
+        }
+
+        // Remove fiducials in the immediate buffer that are no longer in sight
+        auto it = mImmediateFiducials.begin();
+        while (it != mImmediateFiducials.end()) {
+            if (std::find(mIdsCache.begin(), mIdsCache.end(), it->first) == mIdsCache.end()) {
+                it = mImmediateFiducials.erase(it);
+            } else {
+                ++it;
+            }
+        }
+
+        fiducial_msgs::FiducialTransformArray fidArray{};
+        fidArray.header.frame_id = ODOM_FRAME;
+        fidArray.header.stamp = ros::Time::now();
+        fidArray.header.seq = mSeqNum;
+        fidArray.transforms.reserve(mPersistentFiducials.size());
+
+        // Add readings to the persistent representations of the fiducials
+        for (auto [id, immediateFid]: mImmediateFiducials) {
+            PersistentFiducial& fid = mPersistentFiducials[id];
```

Comment:
Was there no expiration logic in the code that we ran at URC for these persistent fiducials? It seems reasonable to have them expire if the drift exceeds a fairly liberal threshold, but I don't want to over-engineer if its not necessary.

Reply:
No there was not. We could do that, but it should remain accurate since the tags are put in the global frame